plusdeck/dbus/client.py

```diff
@@ -0,0 +1,335 @@
+import asyncio
+from collections.abc import AsyncIterable
+from dataclasses import dataclass
+import functools
+import logging
+import sys
+from typing import Any, cast, Optional, Self
+from unittest.mock import Mock
+
+import click
+
+from plusdeck.cli import async_command, AsyncCommand, echo, LogLevel, OutputMode, STATE
+from plusdeck.client import State
+from plusdeck.config import Config
+from plusdeck.dbus.interface import DBUS_NAME, DbusInterface
+
+logger = logging.getLogger(__name__)
+
+
+class DbusClient(DbusInterface):
+    """
+    Plus Deck 2C dbus client.
+    """
+
+    def __init__(self: Self) -> None:
+        client = Mock(name="client", side_effect=NotImplementedError("client"))
+        super().__init__("", client)
+        cast(Any, self)._proxify(DBUS_NAME, "/")
+
+
+@dataclass
+class Obj:
+    """
+    The main click context object. Includes a dbus client and the ability to load
+    the service's active config file.
+    """
+
+    client: DbusClient
+    output: OutputMode
+    _config: Optional[Config] = None
+
+    async def config(self: Self) -> Config:
+        if not self._config:
+            config_file: str = await self.client.config_file
+            config: Config = Config.from_file(config_file)
+            self._config = config
+            return config
+        else:
+            return self._config
+
+
+def pass_config(fn: AsyncCommand) -> AsyncCommand:
```

Comment:
Docstrings in here could use some love. Honestly, it's probably worth reviewing docstrings more generally.